compiler/rustc_codegen_ssa/src/mir/rvalue.rs

```diff
@@ -277,9 +289,49 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
             }
         };
         imm = bx.to_immediate_scalar(imm, to_scalar);
+        self.assume_scalar_range(bx, imm, to_scalar, to_backend_ty);
         imm
     }
 
+    fn assume_scalar_range(
+        &self,
+        bx: &mut Bx,
+        imm: Bx::Value,
+        scalar: abi::Scalar,
+        backend_ty: Bx::Type,
+    ) {
+        if matches!(self.cx.sess().opts.optimize, OptLevel::No | OptLevel::Less)
+            || !matches!(scalar.primitive(), abi::Primitive::Int(..))
```

Comment:
Mostly it's just ease of implementation -- avoiding the need to handle bitcasting floats to integers (and do we ever niche floats to begin with?) or getting the ptrtoint casts in the right places. 

You have a point that it probably ought to handle pointers at some point, since this won't help `*const u32` → `NonNull<u32>`.  I'll add a `FIXME`.